tree_unique_args/src/body_contains.rs

```diff
@@ -0,0 +1,82 @@
+use crate::ir::{Constructor, ESort, Purpose};
+use strum::IntoEnumIterator;
+
+// Builds rules like:
+// (rule ((Let in out))
+//       ((BodyContains (Let in out) out))
+//       :ruleset always-run)
+fn captured_expr_rule_for_ctor(ctor: Constructor) -> Option<String> {
+    let pat = ctor.construct(|field| field.var());
+    let actions = ctor.filter_map_fields(|field| {
+        (field.purpose == Purpose::CapturedExpr)
+            .then(|| format!("(BodyContainsExpr {pat} {e})", e = field.var()))
+    });
+    if actions.is_empty() {
+        None
+    } else {
+        let actions_s = actions.join(" ");
+        Some(format!("(rule ({pat}) ({actions_s}) :ruleset always-run)"))
+    }
+}
+
+// Builds rules like:
+// (rule ((BodyContainsExpr body (Add x y)))
+//       ((BodyContainsExpr body x)
+//        (BodyContainsExpr body y))
+//       :ruleset always-run)
+fn subexpr_rule_for_ctor(ctor: Constructor) -> Option<String> {
+    let pat = ctor.construct(|field| field.var());
+    let actions = ctor.filter_map_fields(|field| {
+        (field.purpose == Purpose::SubExpr || field.purpose == Purpose::SubListExpr).then(|| {
+            format!(
+                "(BodyContains{sort} body {e})",
+                sort = field.sort().name(),
+                e = field.var()
+            )
+        })
+    });
+    if actions.is_empty() {
+        None
+    } else {
+        let actions_s = actions.join(" ");
+        Some(format!(
+            "(rule ((BodyContains{sort} body {pat})) ({actions_s}) :ruleset always-run)",
+            sort = ctor.sort().name()
+        ))
+    }
+}
+
+pub(crate) fn rules() -> Vec<String> {
+    ESort::iter()
+        .map(|sort| "(relation BodyContains* (Expr *))".replace("*", sort.name()))
```

Comment:
nit: why the `replace("*"` instead of `format!("(relation BodyContains{sort} (Expr {sort}))", sort = sort.name())`, lik e in the rest of this file?

Reply:
The honest reason is that so it doesn't wrap across 3 lines